File: quickSort.js
var quickSort = function ( list ) {

  if ( list.length === 0 ) return [];
  
  var pivot = list[0];
  var lesser = [];
  var greater = [];

  for ( var i = 1; i < list.length; i++ ) {
    if ( list[i] < pivot ) {
      lesser.push(list[i]);
    } else {
      greater.push(list[i]);
    }
  }

  return quickSort(lesser).concat(pivot, quickSort(greater));
};

File: quickSort.go
package main

import "fmt"

func quickSort(array []int) []int {

	fmt.Println(array)

	if len(array) < 2 {
		return array
	}

	pivot := array[0]
	left := make([]int, 0)
	right := make([]int, 0)

	for i := 1; i < len(array); i++ {
		if array[i] <= pivot {
			left = append(left, array[i])
		} else {
			right = append(right, array[i])
		}
	}

	fmt.Println("left", left)
	fmt.Println("pivot", pivot)
	fmt.Println("right", right)

	return append(append(quickSort(left), pivot), quickSort(right)...)
}

func main() {
	test := []int{49, 54, 98, 19, 70, 2, 23, 76, 26, 99, 6, 106, 11}
	fmt.Println(quickSort(test))
}
